### ISSUE TITLE
ca65: .include is broken somehow

### ISSUE THREAD

**Issue by kugelfuhr** [cc65/cc65#1655]
I was using a macro in older sources of mine that doesn't work any longer:

```
        .macpack cpu
        .macro  pfinc   what
                .if (.cpu .bitand CPU_ISET_65816)
                        .out .sprintf("Including %s-65816.inc", what)
                        .include .sprintf("%s-65816.inc", what)
                .else
                        .out .sprintf("Including %s-6502.inc", what)
                        .include .sprintf("%s-6502.inc", what)
                .endif
        .endmacro

        pfinc   "math"
```

After some digging, I found PR #1477 and PR #1503 and learned that this was broken deliberately. I don't like it, but probably have to live with it.

I titled this ".include is broken" not because of the change but:

1. The behavior now differs from what is described in the docs. Chapter 12 for example says "If that name is used in the source file, the macro is "expanded", that is, it is replaced by the tokens that were specified when the macro was defined.". This is no longer true. Chapter 10.5 describes how to use `.concat` with `.include`. But this does no longer work in macros. There are probably more places that need to get fixed.
2. The behavior is inconsistent. Inside of a macro, almost none of the builtin functions work together with `.include`. Not `.sprintf`, not `.concat`, not `.left`, `.right`, `.string`, ... Outside of a macro they work. This is bad and software should not work like this. Before the change, all this worked and was consistent. I consider this inconsistency a bug that should be fixed. Until then, all these problems should be documented. Breaking old code is bad enough, so the new solution should at least be ok.

**Comment by mrdudz** [comment]
I have to agree that something that is working outside of a macro should also work inside a macro

**Comment by mrdudz** [comment]
After reading #1477 and #1503 i think i prefer reverting those - fixing one thing while breaking another should never be an option, even more so if it introduces inconsistency like this one.

@ReallyNiceGuy please speak up - are you interested in trying to fix this?

IMHO what we need is some test programs that exercise the old behaviour first. Then we need a test program that tests the supposedly broken things. And only then we can try to fix it. Also the fix must come with an update to the documentation, if the behaviour changes, changing the behaviour of the assembler and not adjusting the documentation is just terrible ;_;

**Comment by ReallyNiceGuy** [comment]
I don't mind you reverting the changes, we seem to have different views on how .include should work. Keeping the older behavior should be preferred, then.

**Comment by mrdudz** [comment]
> we seem to have different views on how .include should work.

no, not at all. However the way it works now changes the behaviour of the assembler inside a macro AND that difference is not documented either. Plus the change breaks perfectly valid code. Like spiro already said in the other discussion, the patch seems to be "too easy". Fixing it to work as you did and not break the pseudo-ops in include lines will require a more sophisticated patch.

**Comment by ReallyNiceGuy** [comment]
Before we can create a more sophisticated patch, we should define what is the *expected* behavior of .include. The old behavior is also glitchy, in my opinion.
But I am not sure if I will be able to do it myself. I have a feeling that, to make .include work as I want it to work, we would need a recursive parser (to resolve names as they are needed, no matter the level of .include or even .printf), but this would be a huge change.

**Comment by mrdudz** [comment]
> Before we can create a more sophisticated patch, we should define what is the expected behavior of .include. The old behavior is also glitchy, in my opinion.

I think first we need to define what we are actually talking about here. It's not the behaviour of .include per se :)

To me, the expected behaviour is:

- a macro is equivalent to copy-pasting the code in the macro into the sourcecode (while substituting the macro arguments of course)
- that means .include happens at the exact position in the macro, and is equal to copypasting the included file into the macro

however, also:

- the argument to .include can be non constant and be constructed from macro arguments and/or pseudo-ops

**Comment by ReallyNiceGuy** [comment]
With the current implementation, it seems you can't have both.

**Comment by mrdudz** [comment]
Sure, that much we know already :) As i said - it will require a more sophisticated patch

**Comment by Movax12** [comment]
Any chance of resolving this? I use `.sprintf` with `.include` in a macro, but I can't use the latest builds with this 'fix', and therefore would have to fork if I implemented new bug fixes or features.

**Comment by ReallyNiceGuy** [comment]
I made a pull request (https://github.com/cc65/cc65/pull/1825) to revert this change. In the mean time, you can try my repository to see if this will solve your issue until the merge is accepted:
https://github.com/ReallyNiceGuy/cc65

**Comment by Movax12** [comment]
I now see the code change is just that one block in the `MacDef` function. For now, I'll just build my own with that commented out. BUT perhaps this could be implemented as a `.feature` ? I'll consider doing so and submitting a PR.

**Comment by mrdudz** [comment]
> BUT perhaps this could be implemented as a .feature ? I'll consider doing so and submitting a PR.

please no - .feature is mostly to simulate other assemblers - not to switch between one quirky way and another.

**Comment by ReallyNiceGuy** [comment]
> > BUT perhaps this could be implemented as a .feature ? I'll consider doing so and submitting a PR.
> 
> please no - .feature is mostly to simulate other assemblers - not to switch between one quirky way and another.

I agree.

**Comment by mrdudz** [comment]
merged #1825 and added a simple test
